src/Services/SearchService.php

```diff
@@ -306,24 +303,57 @@ private function shouldRedirectToProductDetailPage(array $productsIds, HttpReque
     /**
      * @return string|null
      */
-    private function getProductDetailUrl(int $productId)
+    private function getProductDetailUrl(Response $response)
     {
         /** @var ItemSearchService $itemSearchService */
         $itemSearchService = $this->getItemSearchService();
 
+        $productId = $response->getProductsIds()[0];
         $result = $itemSearchService->getResult(
-            $this->getSearchFactory([$productId])
+            $this->getSearchFactory()->hasItemId($productId)
         );
 
         if (!$result['success'] || empty($result['documents'][0])) {
             return null;
         }
 
-        $productData = $result['documents'][0]['data'];
+        $query = $response->getData(Response::DATA_QUERY)['query'];
 
+        $productData = $result['documents'][0]['data'];
         $urlPath = $productData['texts'][0]['urlPath'];
         $itemId = $productData['item']['id'];
+        $variationId = $this->getVariationIdForRedirect($query, $result['documents']);
+
+        return sprintf('/%s_%s_%s', $urlPath, $itemId, $variationId);
+    }
+
+    private function getVariationIdForRedirect(string $query, array $documents)
+    {
+        $query = strtolower($query);
```

Comment:
Try to not override existing variables in the same method. This would be a `$lowercasedQuery`, or you could call the given argument `$rawQuery`, but they should not override eachother.